File: import_workouts.js
#!/usr/bin/env node

const admin = require('firebase-admin');
const fs = require('fs');
const { parse } = require('csv-parse/sync');
const { v4: uuidv4 } = require('uuid');

// Initialize Firebase Admin SDK using Application Default Credentials
// This uses the Firebase CLI's authentication
try {
  admin.initializeApp({
    credential: admin.credential.applicationDefault(),
    projectId: 'burpeebata',
  });
} catch (error) {
  console.error('Error initializing Firebase Admin SDK:', error);
  console.error('\nPlease ensure you are logged in with Firebase CLI:');
  console.error('  firebase login');
  process.exit(1);
}

const db = admin.firestore();
const auth = admin.auth();

// Map CSV burpee types to app enum
function mapBurpeeType(csvType) {
  if (csvType === 'Navy Seal') return 1; // BurpeeType.navySeal
  if (csvType === 'Six Count') return 0; // BurpeeType.militarySixCount
  return 0;
}

// Parse CSV timestamp to Date
function parseTimestamp(timestamp) {
  // Format: "10/27/2025 11:02:22" (MM/DD/YYYY HH:mm:ss)
  const [datePart, timePart] = timestamp.split(' ');
  const [month, day, year] = datePart.split('/');
  const [hours, minutes, seconds] = timePart.split(':');

  return new Date(
    parseInt(year),
    parseInt(month) - 1, // JS months are 0-indexed
    parseInt(day),
    parseInt(hours),
    parseInt(minutes),
    parseInt(seconds)
  );
}

// Convert CSV row to Workout object
function csvRowToWorkout(row) {
  const date = parseTimestamp(row.Timestamp);
  const totalWorkSeconds = parseInt(row.SetCount) * parseInt(row.SetDuration);
  const totalRestSeconds = (parseInt(row.SetCount) - 1) * parseInt(row.RestDuration);
  const elapsedSeconds = totalWorkSeconds + totalRestSeconds;

  return {
    id: uuidv4(),
    date: date.toISOString(),
    burpeeType: mapBurpeeType(row.BurpeeType),
    repsPerSet: parseInt(row.RepsPerSet),
    secondsPerSet: parseInt(row.SetDuration),
    numberOfSets: parseInt(row.SetCount),
    restBetweenSets: parseInt(row.RestDuration),
    completed: row.IsCompleted === 'TRUE',
    completedSets: parseInt(row.SetCount), // Assume all completed if IsCompleted is true
    isCompleted: row.IsCompleted === 'TRUE',
    isCompletedInTime: row.IsOnTime === 'TRUE',
    elapsedSeconds: elapsedSeconds,
  };
}

async function importWorkouts(userId, csvPath) {
  try {
    console.log(`User ID: ${userId}`);

    // Read and parse CSV
    console.log(`Reading CSV file: ${csvPath}...`);
    const csvContent = fs.readFileSync(csvPath, 'utf-8');
    const records = parse(csvContent, {
      columns: true,
      skip_empty_lines: true,
    });
    console.log(`Found ${records.length} workout records`);

    // Convert to workouts
    const workouts = records.map(csvRowToWorkout);

    // Import to Firestore
    console.log(`Importing workouts to Firestore...`);
    const batch = db.batch();
    let count = 0;

    for (const workout of workouts) {
      const docRef = db
        .collection('users')
        .doc(userId)
        .collection('workouts')
        .doc(workout.id);

      batch.set(docRef, workout);
      count++;

      // Firestore batch limit is 500 operations
      if (count % 500 === 0) {
        await batch.commit();
        console.log(`  Committed batch of ${count} workouts...`);
      }
    }

    // Commit remaining
    if (count % 500 !== 0) {
      await batch.commit();
    }

    console.log(`✅ Successfully imported ${workouts.length} workouts`);
    console.log(`📊 Summary:`);
    console.log(`   - User ID: ${userId}`);
    console.log(`   - Total workouts: ${workouts.length}`);
    console.log(`   - Completed: ${workouts.filter(w => w.completed).length}`);
    console.log(`   - Date range: ${workouts[0].date} to ${workouts[workouts.length - 1].date}`);

  } catch (error) {
    console.error('❌ Error importing workouts:', error);
    process.exit(1);
  }
}

// Run the import
const userId = process.argv[2];
const csvPath = process.argv[3] || './user_records_export.csv';

if (!userId) {
  console.error('Usage: node import_workouts.js <userId> [csvPath]');
  console.error('Example: node import_workouts.js 3zqmtJb9CiZxKD7wx8avCGSWHRw1 user_records_export.csv');
  process.exit(1);
}

importWorkouts(userId, csvPath)
  .then(() => process.exit(0))
  .catch(err => {
    console.error(err);
    process.exit(1);
  });

File: import_workouts.py
#!/usr/bin/env python3

import csv
import json
import sys
import requests
from datetime import datetime
from uuid import uuid4

def map_burpee_type(csv_type):
    """Map CSV burpee type to app enum index"""
    return 1 if csv_type == 'Navy Seal' else 0  # 0=militarySixCount, 1=navySeal

def parse_timestamp(timestamp_str):
    """Parse MM/DD/YYYY HH:MM:SS to ISO8601"""
    dt = datetime.strptime(timestamp_str, '%m/%d/%Y %H:%M:%S')
    return dt.isoformat() + 'Z'

def csv_row_to_workout(row):
    """Convert CSV row to Workout document"""
    set_count = int(row['SetCount'])
    set_duration = int(row['SetDuration'])
    rest_duration = int(row['RestDuration'])

    # Calculate elapsed seconds
    elapsed_seconds = (set_count * set_duration) + ((set_count - 1) * rest_duration)

    return {
        'id': str(uuid4()),
        'date': parse_timestamp(row['Timestamp']),
        'burpeeType': map_burpee_type(row['BurpeeType']),
        'repsPerSet': int(row['RepsPerSet']),
        'secondsPerSet': set_duration,
        'numberOfSets': set_count,
        'restBetweenSets': rest_duration,
        'completed': row['IsCompleted'] == 'TRUE',
        'completedSets': set_count,
        'isCompleted': row['IsCompleted'] == 'TRUE',
        'isCompletedInTime': row['IsOnTime'] == 'TRUE',
        'elapsedSeconds': elapsed_seconds,
    }

def import_workouts(user_id, csv_path, project_id='burpeebata'):
    """Import workouts from CSV to Firestore"""

    # Read CSV
    print(f'📄 Reading CSV: {csv_path}')
    workouts = []
    with open(csv_path, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            workout = csv_row_to_workout(row)
            workouts.append(workout)

    print(f'📊 Found {len(workouts)} workout records')
    print(f'👤 Target user ID: {user_id}')
    print(f'🔥 Target project: {project_id}')

    # Import using Firestore REST API
    base_url = f'https://firestore.googleapis.com/v1/projects/{project_id}/databases/(default)/documents'

    print('\n⚠️  To authenticate, you need to provide a Firebase ID token.')
    print('You can get this by:')
    print('1. Sign in to the app as alexandergarber@gmail.com')
    print('2. Get the ID token from Firebase Auth')
    print('\nOr run: gcloud auth print-identity-token')
    print('\nEnter ID token (or press Enter to skip and show cURL commands): ')

    id_token = input().strip()

    if not id_token:
        # Generate cURL commands instead
        print('\n📋 Copy and run these commands to import workouts:\n')
        for i, workout in enumerate(workouts, 1):
            doc_path = f'users/{user_id}/workouts/{workout["id"]}'
            url = f'{base_url}/{doc_path}'

            print(f'# Workout {i}/{len(workouts)}: {workout["date"]}')
            print(f'curl -X PATCH "{url}?updateMask=*" \\')
            print(f'  -H "Content-Type: application/json" \\')
            print(f'  -H "Authorization: Bearer YOUR_ID_TOKEN" \\')
            print(f'  -d \'{json.dumps({"fields": convert_to_firestore_fields(workout)})}\'')
            print()
        return

    # Import with ID token
    headers = {
        'Authorization': f'Bearer {id_token}',
        'Content-Type': 'application/json',
    }

    success_count = 0
    for i, workout in enumerate(workouts, 1):
        doc_path = f'users/{user_id}/workouts/{workout["id"]}'
        url = f'{base_url}/{doc_path}'

        # Convert to Firestore fields format
        data = {'fields': convert_to_firestore_fields(workout)}

        try:
            response = requests.patch(
                url,
                headers=headers,
                params={'updateMask': '*'},
                json=data
            )

            if response.status_code in [200, 201]:
                success_count += 1
                print(f'  ✓ [{i}/{len(workouts)}] Imported: {workout["date"]}')
            else:
                print(f'  ✗ [{i}/{len(workouts)}] Failed: {response.status_code} - {response.text}')
        except Exception as e:
            print(f'  ✗ [{i}/{len(workouts)}] Error: {e}')

    print(f'\n✅ Successfully imported {success_count}/{len(workouts)} workouts!')

def convert_to_firestore_fields(workout):
    """Convert workout dict to Firestore fields format"""
    fields = {}
    for key, value in workout.items():
        if isinstance(value, bool):
            fields[key] = {'booleanValue': value}
        elif isinstance(value, int):
            fields[key] = {'integerValue': str(value)}
        elif isinstance(value, str):
            fields[key] = {'stringValue': value}
    return fields

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print('Usage: python3 import_workouts.py <user_id> <csv_path>')
        print('Example: python3 import_workouts.py abc123xyz user_records_export.csv')
        sys.exit(1)

    user_id = sys.argv[1]
    csv_path = sys.argv[2]

    import_workouts(user_id, csv_path)
